Conceptual: Substrate Storage (#255)
substrate-developer-hub/substrate-developer-hub.github.io: docs/conceptual/core/storage.md

title: Storage
---

The Substrate uses a simple a key-value data store implemented as a database-backed modified Merkle tree.

## Key-Value Database

Substrate implements it's storage database with [RocksDB](https://rocksdb.org/), a persistent key-value store for fast storage environments.

This is used for all the components of Substrate which require persistent storage such as:

* Substrate clients
* Substrate light-clients
* Off-chain workers

## Trie Abstraction

> One advantage of using a simple key-value store is that you are able to easily abstract other storage structures on top.

Reviewer: Can we put a bit more content on _how a Trie is abstracting on top of a K-V store_? I have worked with storage from highest level (SRML), to the lowest (raw RPC to to encoded key) and from the looks of it, honestly, I never really felt like I am working with a Trie here. It always looked more like a bare bone KV store to me.

Author: Would need to know what that information would look like...

Implementation details like this could be above the level of "conceptual" docs, but I agree we would want this information in the reference docs.

Reviewer 2: I would say it is interesting to keep in mind as a runtime developper that the trie is way more costy than a standard k-v store.
General idea being that you should rather store a serialized struct with two field at a storage location rather than both field at different location, except if you want to reduce the size of the proof.
That is something that can get quite clear when you apply storage cost (if you have a  base cost and variable cost for size to store).

Substrate uses a Base-16 Modified Merkle Patricia tree ("trie") from [`paritytech/trie`](https://github.com/paritytech/trie) to provide a trie structure whose contents can be modified and whose root hash is recalculated efficiently.

> Tries are important tool for blockchains because the allows for efficient storing and sharing of historical of block state. You don't have a state trie per block, but a trie hash that will point to the nodes from previous block states. However, accessing data to trie data is costly. Each read operation takes O(log N) time, where N is the number of elements stored in the trie. To mitigate this, we use a key value cache.

Reviewer: ```suggestion
Tries are important tool for blockchains because they allows for efficient storing and sharing of historical of block state. You don't have a state trie per block, but a trie hash that will point to the nodes from previous block states. However, accessing data to trie data is costly. Each read operation takes O(log N) time, where N is the number of elements stored in the trie. To mitigate this, we use a key value cache.
```

Reviewer: And you should use sth like this
https://marketplace.visualstudio.com/items?itemName=stkb.rewrap

## Key-Value Database

Substrate implements its storage database with [RocksDB](https://rocksdb.org/), a persistent key-value store for fast storage environments.

This is used for all the components of Substrate which require persistent storage such as:

* Substrate clients
* Substrate light-clients
* Off-chain workers

## Trie Abstraction

One advantage of using a simple key-value store is that you are able to easily abstract other storage structures on top.

Substrate uses a Base-16 Modified Merkle Patricia tree ("trie") from [`paritytech/trie`](https://github.com/paritytech/trie) to provide a trie structure whose contents can be modified and whose root hash is recalculated efficiently.

Tries are important tool for blockchains because they allow for efficient storing and sharing of historical of block state. Each block does not have it's own state trie, but a trie hash that will point to the trie nodes from previous block states.

Accessing data to trie data is costly. Each read operation takes O(log N) time, where N is the number of elements stored in the trie. To mitigate this, we use a key value cache.

> All trie node are stored in RocksDB and part of the trie state can get pruned, i.e. a key-value pair can be deleted from the storage when it is out of pruning range for non archive nodes. The trie nodes partial path prefixes the hash of the encoded node for the RocksDB key to avoid key collision. We do not use [reference counting](http://en.wikipedia.org/wiki/Reference_counting) for performance reasons.

Reviewer 3: ```suggestion
All trie nodes are stored in RocksDB and part of the trie state can get pruned, i.e. a key-value pair can be deleted from the storage when it is out of pruning range for non archive nodes. The trie nodes partial path prefixes the hash of the encoded node for the RocksDB key to avoid key collision. We do not use [reference counting](http://en.wikipedia.org/wiki/Reference_counting) for performance reasons.
```

title: Storage
---

The Substrate uses a simple a key-value data store implemented as a
database-backed modified Merkle tree.

## Key-Value Database

Substrate implements its storage database with [RocksDB](https://rocksdb.org/),
a persistent key-value store for fast storage environments.

This is used for all the components of Substrate which require persistent
storage such as:

- Substrate clients
- Substrate light-clients
- Off-chain workers

## Trie Abstraction

One advantage of using a simple key-value store is that you are able to easily

> abstract other storage structures on top.

Reviewer 2: ```suggestion
abstract storage structures on top of it.
```
not sure if better, but I first read it as 'abstract over' which got a different meaning.

Substrate uses a Base-16 Modified Merkle Patricia tree ("trie") from
[`paritytech/trie`](https://github.com/paritytech/trie) to provide a trie
structure whose contents can be modified and whose root hash is recalculated
efficiently.

> Tries are important tool for blockchains because they allow for efficient

Reviewer 2: I think 'proof of state inclusion for data' is more important than state history property.
'Tries ar important tool for blockchains because they allow to produce and share small proof of data inclusion in a block. Those data are simple key values. Note that the trie structure also is a simple way to store historical of a block state.' .....

Reviewer 2: In fact just adding the idea of 'provable' should be fine (my formulation is redundant with state trie part).

Reviewer 2: I think I reviewed an old version, looks good here.

storing and sharing of historical of block state. Each block does not have its
own state trie, but a trie hash that will point to the trie nodes from previous
block states.

> Accessing trie data is costly. Each read operation takes O(log N) time,

Reviewer 2: As costy as prooving inclusion of a data (not sure it is interesting through).

---

The Substrate uses a simple a key-value data store implemented as a
database-backed modified Merkle tree.

## Key-Value Database

Substrate implements its storage database with [RocksDB](https://rocksdb.org/),
a persistent key-value store for fast storage environments.

This is used for all the components of Substrate which require persistent
storage such as:

- Substrate clients
- Substrate light-clients
- Off-chain workers

## Trie Abstraction

One advantage of using a simple key-value store is that you are able to easily
abstract other storage structures on top.

Substrate uses a Base-16 Modified Merkle Patricia tree ("trie") from
[`paritytech/trie`](https://github.com/paritytech/trie) to provide a trie
structure whose contents can be modified and whose root hash is recalculated
efficiently.

Tries are important tool for blockchains because they allow for efficient
storing and sharing of the historical block state. The root hash of the trie
provides a cryptographic fingerprint of the final resulting storage after
executing all state transitions by a blockchain. Thus, two blockchain nodes can
easily verify they have the same final state by simply comparing their trie

> root.

Author: updated

Accessing trie data is costly. Each read operation takes O(log N) time, where N
is the number of elements stored in the trie. To mitigate this, we use a key
value cache.

All trie nodes are stored in RocksDB and part of the trie state can get pruned,
i.e. a key-value pair can be deleted from the storage when it is out of pruning

> range for non archive nodes. The trie nodes partial path prefixes the hash of

Reviewer 4: "The trie nodes partial path prefixes"

Super confusing. You were just talking about nodes as nodes in a network. Now, nodes means nodes in the trie.

Something like, "The partial path of a node in the trie is the prefix for..." may be easier to digest.

Author: I just removed this sentence. I think the value add is low. feels like an implementation detail.

the encoded node for the RocksDB key to avoid key collision. We do not use
[reference counting](http://en.wikipedia.org/wiki/Reference_counting) for
performance reasons.

### State Trie

> Substrate has a single main trie, called the state trie, whose changing root

Reviewer 4: "Substrate" does not have a single main trie, "Substrate based chains" do.

"changing" is a poor choice of words here. It's not continuously changing; it changes with discreet events (state updates).

Author: I just removed the adjective

hash is placed in each block header. This is used to easily verify the state of

> the blockchain and provide a basis for light clients to verify proofs.

Reviewer 5: I would be interested to know more about the light clients and what exactly is proved to them. I know they only read block headers (not full blocks) so they get to see the state root before and after each block, but how do they actually know that the transactions in the block were executed correctly and lead to the resulting state root?

Reviewer 2: what is proved to the light client is validity of an operation over a state of the chain. Say the light client does not have the full state, he get the info he need with a proof that it comes from a valid full state.

simplier example should be:

-----------
Querying keys from state trie, the request will be (from cone/network/src/protocol/message.rs):
```
	#[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
	/// Remote storage read request.
	pub struct RemoteReadRequest<H> {
		/// Unique request id.
		pub id: RequestId,
		/// Block at which to perform call.
		pub block: H,
		/// Storage key.
		pub keys: Vec<Vec<u8>>,
	}
```
then the reply will be
```
#[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
/// Remote read response.
pub struct RemoteReadResponse {
	/// Id of a request this response was made for.
	pub id: RequestId,
	/// Read proof.
	pub proof: Vec<Vec<u8>>,
}
```
To make sense out of this reply we fetch request by id and the state trie root of the requested block hash, then we query keys over the proof.
The proof is here a subset of the memorydb the state trie is build upon.
This subset contains only the trie nodes the full client does record when running the query on its side (for every keys from the request).
Then the light client run the same keys query over a trie build from the block state root (he know it from the cht) and the trie nodes he just received (field proof of response). From this incomplete state trie he can get the resulting values of every keys in input (and since every node of trie refers to each over through a crypto hash it is proof it is in the chain state).

------------

but this way of recording some operation on a full client and re-executing over this record on a light client can apply to many thing (if you look in message.rs there is a few query, call for instance uses executor on light blockchain).
Possible future design for substrate light client seems to be allowing evaluation of some wasm code which could be a better solution than chaining queries result like it was done in eth. (better solution to reducing the number of queries roundtrip).

Not 100% sure on the following point, but proving the transition between two consecutive blocks is not something we do (I think), we just rely on the fact that the blocks got validated and we know they are chained. Surely a full client can execute the transition by running a full block on its state, so it should be possible for him to send to a light client all accessed db keys during the full block execution and next root calculation, and the light client will be able to execute the block on those keys and produce next root. But I guess the proof will be really massive, so relying on network having validating those state may be enough in the light client usecase.

But you can run any call already (see `RemoteCallRequest`) or query the deltas between block through `RemoteChangesRequest`.

Author: I think a slightly simpler, but hopefully still fully accurate explanation of light client proofs come from merkle proof in general.

![image](https://user-images.githubusercontent.com/1860335/65807398-85542280-e18e-11e9-88a1-3f3e6844db42.png)

from here: https://www.quora.com/Cryptography-How-does-a-Merkle-proof-actually-work

The image hopefully shows that if you want to prove TX3 is in the trie, you dont need ALL the nodes, just the full branch will leads to TX3 and all the nodes next to the nodes on that full branch. Obviously WAYWAY less data on big tries like the ones on blockchains.

This trie only stores content for the canonical chain, not forks. There is a
separate `state_db` layer that maintain the trie state with references counted
in memory for all that is non canonical. See JournalDB.

### Child Trie

Substrate also provides an API to generate new child tries with their own root
hash that can be used in the runtime.

Child tries are identical to main state trie, except their root is stored and
updated in the main trie instead of the block header. Since their headers are a
part of the main state trie, it is still easy to verify of the complete node
state when it includes child tries.

## Runtime Storage API

> The Substrate runtime support library provides utilities which generates unique,

Reviewer 4: ```suggestion
The Substrate runtime support library provides utilities to generate unique,
```

a persistent key-value store for fast storage environments.

This is used for all the components of Substrate that require persistent storage
such as:

- Substrate clients
- Substrate light-clients
- Off-chain workers

## Trie Abstraction

One advantage of using a simple key-value store is that you are able to easily
abstract storage structures on top of it.

Substrate uses a Base-16 Modified Merkle Patricia tree ("trie") from
[`paritytech/trie`](https://github.com/paritytech/trie) to provide a trie
structure whose contents can be modified and whose root hash is recalculated
efficiently.

Tries allow efficient storing and sharing of the historical block state. The
trie root is a representation of the data within the trie; that is, two tries
with different data will always have different roots. Thus, two blockchain nodes
can easily verify they have the same final state by simply comparing their trie
root.

Accessing trie data is costly. Each read operation takes O(log N) time, where N
is the number of elements stored in the trie. To mitigate this, we use a key
value cache.

All trie nodes are stored in RocksDB and part of the trie state can get pruned,

> i.e. a key-value pair can be deleted from the storage when it is out of pruning

Author: There is some configuration, even a CLI flag for Substrate nodes to define the pruning range. Basically the number of blocks to hold data for before we start pruning things in the trie. An archive node has a pruning range of inifinity.

reasons.

### State Trie

Substrate based chains have a single main trie, called the state trie, whose
root hash is placed in each block header. This is used to easily verify the
state of the blockchain and provide a basis for light clients to verify proofs.

This trie only stores content for the canonical chain, not forks. There is a
separate `state_db` layer that maintain the trie state with references counted
in memory for all that is non canonical. See JournalDB.

### Child Trie

Substrate also provides an API to generate new child tries with their own root
hash that can be used in the runtime.

Child tries are identical to main state trie, except their root is stored and
updated in the main trie instead of the block header. Since their headers are a
part of the main state trie, it is still easy to verify the complete node state
when it includes child tries.

Child tries are useful when you want your own independent trie with a separate
root hash that you can use to verify the specific content in that trie.
Subsections of a trie do not have a root-hash-like representation that satisfy
these needs automatically; thus a child trie is used instead.

## Runtime Storage API

> The Substrate runtime support library provides utilities to generate unique,

Author: yup

deterministic keys for your runtime module storage items. These storage items
are placed in the state trie and are accessible by querying the trie by key.

## Next Steps

### Learn More

- Learn how to add [storage items](development/module/storage.md) into your
  Substrate runtime modules.

### Examples

> - View an example of creating [child tries]() in your Substrate runtime module.

Author: needs to get created, I think we will do a second pass at the whole docs things once all the pages make it in to buff out the next steps content.

Substrate uses a Base-16 Modified Merkle Patricia tree ("trie") from [`paritytech/trie`](https://github.com/paritytech/trie) to provide a trie structure whose contents can be modified and whose root hash is recalculated efficiently.

Tries are important tool for blockchains because the allows for efficient storing and sharing of historical of block state. You don't have a state trie per block, but a trie hash that will point to the nodes from previous block states. However, accessing data to trie data is costly. Each read operation takes O(log N) time, where N is the number of elements stored in the trie. To mitigate this, we use a key value cache.

> All trie node are stored in RocksDB and part of the trie state can get pruned, i.e. a key-value pair can be deleted from the storage when it is out of pruning range for non archive nodes. Trie nodes are also encoded to allow for any key type, while still be able to avoid key collision. We do not use [reference counting](http://en.wikipedia.org/wiki/Reference_counting) for performance reasons.

Reviewer 2: Second sentence is a bit unclear, I would rephrase with:
"Trie nodes partial path is also prefixing the hash of encoded node in RocskDb keys, to avoid key collision."
May be a bit to technical (it assumes that readers know about traditional key construction (hash of the encoded node).
